### ISSUE TITLE
Faster startup -- Experiment C -- lazy unmarshalling

### ISSUE THREAD

**Issue by gvanrossum** [faster-cpython/ideas#66]
This is a placeholder for a description of Experiment C (see https://github.com/faster-cpython/ideas/issues/32#issuecomment-878517764)

- borrows ideas from (#64): delay hydrating code objects until called, keep PYC file in memory
- but dehydration just calls back into marshal
- this might be less code changes compared to Experiment A, and doesn't need as many new opcodes

**Comment by gvanrossum** [comment]
- In marshal.loads(), when r_object() encounters the type code TYPE_CODE, instead of constructing a full code object, it should construct a dummy ("dehydrated") code object and skip to the end of the data for that type code.
- The dehydrated code object must be populated with just enough things so that we can construct a function object from it, and introspecting the function will function correctly.
- When we are asked to execute a dehydrated code object, we go back to marshal.loads() and load the full code object (but leave nested code objects as dummies again). This process is called hydration.
- We need one modification to the marshal format: we need to know the size of the code object so that we can skip it. This is harder than it might seem, since writing a marshal file is a one-pass process, and the size is not known until we're done writing the object.
  - A hacky approach would be to have a magic string that we hope never occurs, so we can search for it. Maybe a unique random 8-byte string would suffice. A complication is that code objects nest; we could use different strings for different levels. Another complication is that if the magic string is included anywhere in the data (e.g. in a bytes constant) this would break the format. For a quick experiment this might be acceptable.
  - Another approach would be to marshal to a temporary buffer, once done measure the buffer, and copy the buffer, prefixing it with its size as a 4-byte int. This is slow because of the extra copying (and, again, the nesting makes it worse).
  - If we can update bytes earlier in the output string, this can be with less overhead by simply reserving 4 bytes and patching in the size once done writing. For marshal.dump() (which writes to a file) this is not possible, but for marshal.dumps() it should be simple. Fortunately because of importlib, we always use marshal.dumps().

**Comment by iritkatriel** [comment]
What if marshal loads reads the data in the opposite direction than dumps writes?

**Comment by gvanrossum** [comment]
You mean reading the file backwards? We'd have to change the entire format.

**Comment by iritkatriel** [comment]
Yes.

**Comment by gvanrossum** [comment]
So we'd write the file in reverse polish notation. Unfortunately buffered I/O APIs aren't really geared to reading backwards. I think I'll do this:
- If writing to a string, use the backpatch trick
- Otherwise, code objects will be loaded eagerly
This is all academic since real PYC files always use dumps()/loads(), see the code in `_bootstrap_external.py`.

**Comment by gvanrossum** [comment]
The first step (prefixing the marshalled code object with its size in bytes) is complete in https://github.com/gvanrossum/cpython/tree/lazy-marshal .

**Comment by gvanrossum** [comment]
Let's see what the next step will be. In Experiment A (#64) I already wrote code for creating dehydrated code objects and deciding when to hydrate them, I could probably just copy that. But the implementation of hydration will be different, so let me focus on how to do that.

I worry most about the "refs" feature in the marshal format. Unfortunately the format is not documented except in the code, so I'll just have to refer to that. The basic idea is simple enough:

- When marshal.loads() encounters a top-level code object, it should unmarshal that code object as before. But when it encounters a code object nested inside another, it should construct a dummy code object and skip the data, using the skip size that I added in the previous step. (Conveniently, there's a 'depth' field in the RFILE struct.) The dummy code object must contain a a pointer to the bytes object from which it is reading, plus a pointer to the starting point. (Alternatively, we could make a slice of the original bytes so that the start is at index zero, but we're trying to save time here, so let's not make redundant copies.)
- The bytes object is not accessible from the RFILE; we probably have to add it -- or maybe we should add the Py_buffer struct (ownership TBD). The general approach here can be similar to what I did for #64 -- introduce a new struct containing everything that's needed and that's shared between all code objects for a given PYC file (each of which points to it), and give each code object one extra piece of data needed for hydration -- that could be the offset relative to the start of the buffer where the code object's data starts.
- The tricky bit is the 'refs' field of RFILE. This is a list of certain previously encountered values, used to save space in the marshalled data. For example, if we marshal the tuple ("abc", "abc"), the generated marshal code writes a type code with a flag bit set for the first occurrence of "abc", and instead of the second occurrence (or further) of the same value, it generates a TYPE_REF type code followed by the index of the first value in the list of shared values. (Which happens to be 1, because the tuple value itself also gets shared, even though it occurs only once. This could be improved upon, at the cost of more data analysis in marshal.dumps().)
- When the reader skips over a code object, it will potentially also skip over new values to be entered into the list of shared values. This is a problem for us, since a value shared between two code objects will be entered into the list by the first code object and referenced by the second. If the second code object is hydrated first, the referenced value will be missing! Even more disturbing, if we skip unmarshalling some reference definitions, the indexes of further reference definitions will be off.
- For a quick prototype we could probably get away with disabling the creation of reference definitions, but I need to know first that we could solve this if we had to, because the savings by not repeating identical values is potentially significant (though it isn't really for this toy example).
- To avoid the indexes from getting out of sync, we could add a second new field for marshalled code objects, after the "skip bytes" field, to indicate how many reference definitions were skipped. On unmarshalling, we could then append that many None values to the 'refs' list (fortunately, None has its own type code, TYPE_NONE, so None cannot be a valid value in the list). When rehydrating, we can patch those Nones with the corresponding correct values.
- But what to do if a reference points to a None that hasn't been patched yet? Maybe instead of None, we could put a special object in the list that means "a value needs to be patched here" which has an extra field meaning "here's where you have to start reading to get the missing value". Honestly the simplest thing to put here would perhaps be the dummy code object, since we also don't want to have to unmarshal that code object more than once (once when the first missing value is needed, once when rehydrating the code object). It shouldn't be hard to arrange that *all* code objects in the list of refs are such "to be rehydrated" dummy code objects, since in normal usage code objects are never shared.

Let me do some experiments here before rambling on.

**Comment by gvanrossum** [comment]
I believe I can make this work, although I need to think more about what to do when a value needed requires recursively hydrating several code objects. Honestly, the quickest way to *some* kind of working code is to just disable shared references. (Since so many constants are simple, having MAKE_INT and LOAD_COMMON_CONSTANT might be sufficient. See Experiment B, #64.)

**Comment by gvanrossum** [comment]
Thinking about the design for the rehydration, let's say marshal.c exports a function Hydrate(code) which takes a dehydrated code object and hydrates it. How does it do that?
- The code object must contain a pointer to a struct that tells it (a) the bytes object from which to load, and (b) where in that bytes object to start. (If that pointer is NULL the code object is already hydrated and we do nothing.)
- That pointer is put there by the initial marshal.loads() call that returned the code object in dehydrated form. The pointer must be shareable between all code objects loaded from the same bytes object.
- Who owns the struct? In the prototype, it's immortal: it's allocated by malloc() and never freed. Eventually it should become a PyObject and will be DECREF'ed when a PyCodeObject pointing to it is deallocated.
- The bytes object is not currently a part of the RFILE struct. We'll have to add it there.
- It doesn't even have to be a bytes object (it could be anything with a buffer interface, PEP 3118). But that seems overkill so I'll punt if it isn't.
- Let's call this the **context struct** for now.
- Note that the offset where a particular code object lives is not part of the context struct -- it is a second new field in the code object.

Okay, so now we have a bytes object and a starting point and a code object. What do we do?
- Let's assume the starting point is the index of the TYPE_CODE type-code in the bytes object. We can assert that the first byte here is TYPE_CODE.
- We construct an RFILE struct from the context struct.
- We collect all the data for the code object but instead of constructing a new code object, we somehow update the existing object in place.
- Wish we could use the replace() method on the code object, but it's probably easier (from C code) to add a _PyCode_Update() method that takes a code object and a code constructor and updates the code object in place -- it can share code with _PyCode_New() (maybe a little refactoring is needed).

For code objects containing other code objects (modules, classes, and functions containing lambdas or nested functions) as long as the outer object is not hydrated the inner objects don't exist at all (since they would live in co_consts which is NULL for a dehydrated code object). So these inner code object are created (one level deep) when the outer code object is hydrated. It is important that they all share the same context struct. (So maybe the RFILE has to have an optional pointer to the context struct.)

**Idea for the context struct:** Maybe marshal_loads_impl() can allocate the context struct (if applicable) and add it to the RFILE struct.

In the future, if/when we restore the 'refs' mechanism, we'll have to update this design.

**Comment by iritkatriel** [comment]
> * The code object must contain a pointer to a struct that tells it (a) the bytes object from which to load, and (b) where in that bytes object to start. (If that pointer is NULL the code object is already hydrated and we do nothing.)

Why is this only for code objects and not all marshalled objects? I think the refs could be to any object.

**Comment by gvanrossum** [comment]
> Why is this only for code objects and not all marshalled objects? I think the refs could be to any object.

We only leave code objects potentially dehydrated.

The refs can indeed be any object -- this is one of the problems I'm working around by disabling them for now, but we can't very well add fields for hydration support to e.g. tuples or strings.

**Comment by gvanrossum** [comment]
The prototype in the branch "works" on a very simple test. It needs tests and perf measurements, and there's coding work on memory deallocation, plus the 'refs' list (the latter is the only thing that needs deep thinking).

**Comment by gvanrossum** [comment]
Stuff to do includes:
- [ ] Avoid leaking values in _PyCode_Update()
- [ ] Properly free 'struct context' (by making it an object?)
- [ ] Unit tests
- [ ] Perf measurements
- [ ] Hydrate in more places, e.g. when accessing attributes
- [ ] And the big one: implement 'refs' list again (I need to draw some pictures)

**Comment by gvanrossum** [comment]
So here are the first perf measurements, using Irit's script from the pyco branch (with small tweaks).
```
                                                        load+exec   steady state
shared call locals  consts  100 funcs,   10 vars            1.153          1.091
unique call locals          100 funcs,   10 vars            1.148          0.897
shared call locals          100 funcs,   10 vars            1.092          0.990
shared call globals consts  100 funcs,   10 vars            1.084          1.015
unique call locals  consts  100 funcs,   10 vars            1.075          0.895
shared call globals         100 funcs,   10 vars            1.074          1.029
unique call globals consts  100 funcs,   10 vars            1.055          0.929
unique call globals         100 funcs,  100 vars            1.023          0.996
unique call globals consts  100 funcs,  100 vars            1.022          0.963
shared call globals consts  100 funcs,  100 vars            1.015          1.023
shared call locals          100 funcs,  100 vars            1.011          0.939
unique call locals  consts  100 funcs,  100 vars            1.004          1.026
shared call locals  consts  100 funcs,  100 vars            1.003          0.976
shared call globals         100 funcs,  100 vars            0.998          1.014
unique call locals          100 funcs,  100 vars            0.993          0.992
unique call globals         100 funcs,   10 vars            0.922          1.071
unique      globals consts  100 funcs,   10 vars            0.670          1.004
unique      globals         100 funcs,   10 vars            0.637          0.979
unique      globals consts  100 funcs,  100 vars            0.629          1.095
unique      globals         100 funcs,  100 vars            0.614          1.114
shared      globals consts  100 funcs,   10 vars            0.453          1.130
shared      globals         100 funcs,   10 vars            0.450          0.627
shared      locals          100 funcs,   10 vars            0.337          0.870
shared      locals  consts  100 funcs,   10 vars            0.321          0.833
unique      locals          100 funcs,   10 vars            0.317          0.813
unique      locals  consts  100 funcs,   10 vars            0.292          1.188
shared      globals consts  100 funcs,  100 vars            0.243          1.132
shared      globals         100 funcs,  100 vars            0.242          1.020
shared      locals  consts  100 funcs,  100 vars            0.057          0.768
shared      locals          100 funcs,  100 vars            0.056          0.863
unique      locals          100 funcs,  100 vars            0.049          0.956
unique      locals  consts  100 funcs,  100 vars            0.044          0.894
```

**Comment by gvanrossum** [comment]
Interpretation:
- The variation in "steady state" just shows how fickle running timing tests on my (Windows) development laptop is.
- As with pyco, "unique locals consts 100 f, 100 v" is the fastest. But where in pyco it was a modest 0.23 (~4x faster), here it's a whopping 0.044 (22x faster)!
- Of the non-call lines, the slowest here is 0.67, vs. in pyco the slowest line (a different one) is 2.41.
- The call lines vary from 0.92 to 1.15, which is fairly close around 1.0, i.e. a lazy loads + hydration is not much slower (or faster) than an eager loads. It shouldn't be faster, that's just test noise again.

Oh, FWIW, here are the **pyco numbers** I used for comparison, measured in the [pyco branch](https://github.com/gvanrossum/cpython/tree/pyco) using Tools/pyco/perf_micro.py there. (The call numbers are missing due to a memory leak.)
```
                                                        load+exec   steady state
 shared     globals         100 funcs,  100 vars            2.414          1.076
 shared     globals consts  100 funcs,  100 vars            2.352          1.163
 shared     globals consts  100 funcs,   10 vars            2.210          1.030
 shared     locals          100 funcs,  100 vars            2.043          0.987
 shared     locals  consts  100 funcs,  100 vars            2.026          1.023
 shared     globals         100 funcs,   10 vars            1.898          1.008
 shared     locals          100 funcs,   10 vars            1.507          0.345
 shared     locals  consts  100 funcs,   10 vars            1.443          0.435
 unique     globals         100 funcs,   10 vars            1.355          1.050
 unique     globals         100 funcs,  100 vars            1.286          1.133
 unique     globals consts  100 funcs,   10 vars            1.207          1.057
 unique     globals consts  100 funcs,  100 vars            1.109          1.160
 unique     locals          100 funcs,   10 vars            0.603          0.749
 unique     locals  consts  100 funcs,   10 vars            0.529          1.508
 unique     locals          100 funcs,  100 vars            0.250          1.004
 unique     locals  consts  100 funcs,  100 vars            0.231          1.071
```

**Comment by gvanrossum** [comment]
(Note that this just compares eager and lazy loads *with the same executable*. We should separately compare eager loads with this code in place to loads before I made these changes.)

**Comment by gvanrossum** [comment]
Okay, the algorithm for 'refs' seems simple enough.

Recall that the basic writing algorithm puts a flag bit on some objects' type code in the serialized data stream, which signals to the reader to save that object in a list (named 'refs'). Later items in the data stream can then reference those items by a special TYPE_REF type code, which takes an index in that list.

When the lazy reader skip a code object in the stream, it produces a dehydrated code object instead. The skip information includes a count of the number of list items to skip in 'refs'. *The lazy reader stuffs the list with references to the dehydrated code object.* (The assumption here is that code objects never need to appear "natively" in the 'refs' list, because the compiler never generates multiple references to the same code object. There's a rare exception to this, which we will have to rule out explicitly -- two identical lambdas on the same line share a code object.)

When a reference to a skipped item in 'refs' is encountered, it is recognized as a skipped item by being a dehydrated code object. Rather than trying to be clever, at this point we just hydrate that object, which should replace the skipped item markers with the actual values.

It is possible that a skipped item is actually defined inside a nested code object. Hydrating the outer object will then construct a dehydrated code object for the inner code object, and the items in 'refs' skipped on behalf of that inner code object will be replaced with references to the inner code object.

Therefore, the code to find the actual value of a reference is something like
```c
PyObject *get_ref_value(list refs, int i)
{
    while (refs[i] is a dehydrated code object) {
        hydrate(refs[i]);
    }
    return refs[i];
}
```
Two concerns here:
- If references across code objects are too common we lose a lot of the advantage of hydration. (We might limit the use of references in this case.)
- We need to ensure that the code is thread-safe. (In general marshal is protected by the GIL, but there are places where it may be released, in particular the audit operation. We ought to ensure that if two threads attempt to hydrate the same code object they cannot corrupt any data.)

**Comment by gvanrossum** [comment]
BTW, this approach doesn't help much for startup time of a named interpreter (`python -S -c pass`), because the modules that are imported during startup contain very little unused code. But it ought to help with startup time of scripts that import a lot of stuff without using most of it.

**Comment by gvanrossum** [comment]
Here's a PNG with a more visual explanation. (Sorry, it's very tall!)

![Lazy-marshal](https://user-images.githubusercontent.com/2894642/126018775-2b337582-d53b-46ad-8e06-85d6da67bf8a.png)

**Comment by gvanrossum** [comment]
(Once again, Microsoft tools disapppoint me. I can save either as a low-res PNG or as SVG which somehow is seen as invalid XML. Maybe it's still readable.)

**Comment by iritkatriel** [comment]
I can't quite make out sections 1-3, the text is too small.

**Comment by gvanrossum** [comment]
Here's another version, I moved the text blocks around, but MS Whiteboard doesn't seem to have a way to scale a selection, so the text of (1), (2) and (3) is still much smaller than that of (4) and (5). (This originated in me accidentally changing the magnification factor as I was drawing/writing.)

If you zoom in you should be able to read the smaller text though.

![Lazy-marshal-wide](https://user-images.githubusercontent.com/2894642/126205990-68cf4fc3-ce41-451f-8477-0fb13c5856a9.png)

**Comment by iritkatriel** [comment]
That's clear, thanks.

**Comment by iritkatriel** [comment]
When you do 
```
    while (refs[i] is a dehydrated code object) {
        hydrate(refs[i]);
    }
```
The code object you hydrate could contain more refs to other subtrees. So this needs to be able to handle that.

**Comment by gvanrossum** [comment]
Dang. I knew there was something I was missing. I've mostly been chasing Heisenbugs, probably due to not changing the magic number often enough, which didn't leave enough time to actually think deeply enough.

Those other subtrees will try to resolve their references recursively. But that could cause multiple attempts to hydrate the same code object, and in the worst case infinite recursion.

We'd need to use a queue of sorts to resolve this, right? Or an explicit stack? That's a pain. Things are so much simpler without references -- (1) and (3). Maybe references could be limited to within the same code object (so subconstants may still be shared), and string sharing could be done via interning.

**Comment by iritkatriel** [comment]
> We'd need to use a queue of sorts to resolve this, right? Or an explicit stack? That's a pain.

Could it not be just an "I am already hydrating" flag on the object?

**Comment by iritkatriel** [comment]
Though, it might actually not be that complex because refs only point to the left. So it's quite structured.

**Comment by iritkatriel** [comment]
I think it can work with a little care. The hydration of a tree is in DFS pre-order where siblings are visited from left to right. Then if a ref is visited, it points to the left so if it's in the subtree that we are hydrating then it has already been hydrated. If it's to the left of this subtree, then the recursion into that node's subtree is independent of where we are now (if it was not, then our subtree would have been rooted higher than it is). So no issues.

**Comment by gvanrossum** [comment]
Thanks! That sounds right. Now I have to figure out why sometimes my code crashes.

I'll push what I have to my branch regardless.

**Comment by gvanrossum** [comment]
So far the only crashes I still see are due to stack overflow in the debug build on Windows, which happens only because I turned of PY_LOCAL_AGGRESSIVE in debug mode so I could debug some code in ceval.c. But that part has been debugged, so I just pushed a revert of that.

So far there's one legitimate failure: test_Module in test_marshal.py.

The trick of using a dehydrated code object to mark positions in 'refs' that haven't been hydrated yet fails here. This test takes a code object 'c' and then makes a tuple '(c, c)', then verifies (somewhat indirectly) that after marshal+unmarshal it is still a tuple whose items are the same object.

This is now the only test in test_marshal.py that fails. Before I tried to implement 'refs' handling there were failing 5-10 tests there, all due to various object sharing semantics being broken.

Still waiting for the full test suite to pass, but I'm hopeful now.

**Comment by gvanrossum** [comment]
So a fix for the (c, c) test would probably require a second, parallel array containing the dehydrated code object "owning" missing references. That would be a bit of a shame but not the end of the world.

I'm actually more worried about the performance impact the hydration loop in reference resolution. If there are a lot of references that require hydrating other code objects, the savings due to leaving code objects dehydrated are reduced.

Our simple benchmarks probably won't show this, but in real code it seems to occur a lot (based on a simple printf() call I added).

**Comment by gvanrossum** [comment]
A more pressing issue is that functions created from dehydrated code objects don't have a docstring! This is because the docstring is stored as co_consts[0]. See also https://bugs.python.org/issue36521. I expect that the best solution here is to store the docstring as a separate co_doc field on the code object (though this won't necessarily make Raymond happy). This causes so many test failures when run with LAZY=1 that I need to make this my priority.

**Comment by iritkatriel** [comment]
You can't just hydrate the object when the doc string is needed?

**Comment by gvanrossum** [comment]
> You can't just hydrate the object when the doc string is needed?

I could, but that would means that constructing a function object from a code object requires immediately hydrating the code object -- the very thing we're trying to avoid!

We could make __doc__ on a function a property that hydrates the code object if needed. Setting it would overwrite the func_doc field and if that's not NULL it would override the code object's co_doc.

That solution is actually better than introducing co_doc and putting that early in the marshal data (with co_name, co_qualname and co_filename), because it avoids copying the data of the docstring eagerly. In fact, we might be able to do this for func_name and func_qualname as well, and then we don't have to unmarshal those fields eagerly. And co_filename isn't even needed by the function object, so we could load that lazily as well.

We might have to hydrate code objects in a few other places to make all this work seamlessly (e.g. code_repr() uses co_name and co_filename), but that shouldn't be so bad. (The only risk would be a C stack overflow if you're calling code_repr() when the C stack is already close to full, since unmarshalling could run quite a few stack levels deep. Another reason to implement at least part of PEP 651.)

**Comment by gvanrossum** [comment]
(Done.)

**Comment by iritkatriel** [comment]
I think "unique globals" is the worst micro-benchmark now, showing about 35% slowdown for load+exec. You need to increase num_funcs to see it:

+        [1000],  # num_funcs
+        [500],  # func_length
+        [10],  # num_vars
+        [False],  # is_locals
+        [True],  # is_unique_names
+        [False],  # is_vary_constants
+        [False],  # is_call

**Comment by gvanrossum** [comment]
Yeah, I see that too. I don't have an explanation yet.

I do have an unrelated idea. There are cases where e.g. the names in a code object require hydrating some nested code object. For example:
```
def outer():
    print("outer")
    def inner():
        print("inner")
```
Both outer and inner refer to "print", and it so happens that outer's co_consts is serialized before its co_names, so that the "print" reference in outer's names is a TYPE_REF pointing to the "print" definition in inner.

We can fix this case by changing the order in which the parts of a code object are serialized, in particular by putting consts last. (A further optimization would be to reorder the items in co_consts so that code objects come last, but that requires a change to the compiler, so I'm putting that off.)

I have no expectation that changing this will fix that benchmark, but it will at least reduce one source of premature hydration.

**Comment by iritkatriel** [comment]
> I have no expectation that changing this will fix that benchmark, but it will at least reduce one source of premature hydration.

I think this is related to the benchmark. I saw in the debugger that the elements of the names tuple of each function were not finding themselves in the hashtable so they were registering themselves as refs. I guess later when the module's names (the global) are serialized they reference the names from the nested functions.

With your change the load+exec time looks good now (21% faster on the mac), but in steady-state we are 39% slower. I can't see why, but we are not hydrating code objects anymore.

**Comment by gvanrossum** [comment]
I can't explain it. It seems pretty consistent though (I see it on Mac and Windows and regardless of some other stuff).

I put print calls in (and reduced the numbers in the benchmark), and only the toplevel code object is being hydrated.

(PS. There is still one hard-coded constant in the script -- it does 1000 repetitions of the marshal.loads() call.)

**Comment by gvanrossum** [comment]
Only thing I know for sure is that all Lazy execs were faster before I added refs support back, but Classic load is now 4x faster (I guess the sharing of common subtrees *really* makes a difference for this benchmark).

Compared to main, classic times have not budged (not much anyway) so somehow Lazy exec really got slower?!

**Comment by gvanrossum** [comment]
I went over the changes since 2e2d4c3059 carefully, and maybe it's this loop in r_object()?
```
+                while (p->refs_pos < save_refs_pos + nrefs) {
+                    Py_ssize_t ref = r_ref_reserve(FLAG_REF, p);
+                    if (first_ref < 0)
+                        first_ref = ref;
+                }
```
Combined with this:
```
+            if (first_ref >= 0) {
+                // Overwrite skipped references with v
+                assert(p->ctx != NULL && p->ctx->code == NULL);
+                while (first_ref < p->refs_pos) {
+                    r_ref_insert(v, first_ref++, FLAG_REF, p);
+                }
+            }
```

**Comment by iritkatriel** [comment]
That loop is probably not the issue, it doesn't run after the first exec.

**Comment by iritkatriel** [comment]
In lazy_marshal_perf.py, if I reverse the order so that we calculate new_timings first and classic_timings second, the result changes quite dramatically (classic exec becomes slower):
```
cpython> .\python.bat .\Tools\scripts\lazy_marshal_perf.py
Running Release|x64 interpreter...

Starting speed test: unique      globals        1000 funcs,   10 vars
Classic load: 4.433
Classic exec #1: 1.690
Classic exec #2: 1.647
Classic exec #3: 1.599
Classic exec #4: 1.585
       Classic total: 10.956
Lazy load: 0.258
Lazy exec #1: 4.445
Lazy exec #2: 2.088
Lazy exec #3: 2.025
Lazy exec #4: 2.105
       Lazy total: 10.922
Classic-to-new comparison:
>> load+exec ratio: 0.77 (new is 30% faster)
>> steady-state ratio: 1.33 (new is -25% faster)
```

```
cpython> git diff .\Tools\
diff --git a/Tools/scripts/lazy_marshal_perf.py b/Tools/scripts/lazy_marshal_perf.py
index e1ad72bd20..111d42952b 100644
--- a/Tools/scripts/lazy_marshal_perf.py
+++ b/Tools/scripts/lazy_marshal_perf.py
@@ -40,9 +40,9 @@ def helper(data, label, lazy):

     code = compile(source, "<old>", "exec")
     data = marshal.dumps(code)
-    classic_timings = helper(data, "Classic", lazy=False)

     new_timings = helper(data, "Lazy", lazy=True)
+    classic_timings = helper(data, "Classic", lazy=False)

     if classic_timings and new_timings:
```
```

cpython> .\python.bat .\Tools\scripts\lazy_marshal_perf.py
Running Release|x64 interpreter...

Starting speed test: unique      globals        1000 funcs,   10 vars
Lazy load: 0.242
Lazy exec #1: 4.302
Lazy exec #2: 2.058
Lazy exec #3: 2.064
Lazy exec #4: 2.073
       Lazy total: 10.739
Classic load: 4.181
Classic exec #1: 2.303
Classic exec #2: 2.285
Classic exec #3: 2.210
Classic exec #4: 2.273
       Classic total: 13.253
Classic-to-new comparison:
>> load+exec ratio: 0.70 (new is 43% faster)
>> steady-state ratio: 0.91 (new is 10% faster)

 
```

**Comment by iritkatriel** [comment]
I refactored it to split load from exec and it seems to be the load stage of lazy which causes the slowdown of classic exec. Could this be related to memory alignments or something like that?

```
Starting speed test: unique      globals        1000 funcs,   10 vars
Lazy load: 0.244
Classic load: 0.240
Classic exec #1: 5.077
Classic exec #2: 2.799
Classic exec #3: 2.800
Classic exec #4: 2.801
Lazy exec #1: 4.984
Lazy exec #2: 2.951
Lazy exec #3: 2.924
Lazy exec #4: 2.734
Classic-to-new comparison:
>> load+exec ratio: 0.98 (new is 2% faster)
>> steady-state ratio: 0.98 (new is 2% faster)
```

Vs.

```
Starting speed test: unique      globals        1000 funcs,   10 vars
Classic load: 0.247
Classic exec #1: 4.871
Classic exec #2: 2.083
Classic exec #3: 2.181
Classic exec #4: 2.137
Lazy load: 0.235
Lazy exec #1: 5.094
Lazy exec #2: 2.870
Lazy exec #3: 2.850
Lazy exec #4: 2.757
Classic-to-new comparison:
>> load+exec ratio: 1.04 (new is -4% faster)
>> steady-state ratio: 1.29 (new is -22% faster)
```


```
@@ -17,16 +17,18 @@ def speed_comparison(source: str, test_name: str):
     print()
     print(f"Starting speed test: {test_name}")

-    def helper(data, label, lazy):
-        timings = {}
-        t0 = time.perf_counter()
+    def load_helper(data, label, timings, lazy):
         codes = []
+        t0 = time.perf_counter()
         for _ in range(1000):
             code = marshal.loads(data, lazy=lazy)
             codes.append(code)
         t1 = time.perf_counter()
         print(f"{label} load: {t1-t0:.3f}")
         timings["load"] = t1 - t0
+        return codes
+
+    def exec_helper(codes, label, timings):
         timings["execs"] = []
         for i in range(4):
             t3 = time.perf_counter()
@@ -35,14 +37,27 @@ def helper(data, label, lazy):
             t4 = time.perf_counter()
             print(f"{label} exec #{i+1}: {t4-t3:.3f}")
             timings["execs"].append(t4 - t3)
+
+    def helper(data, label, timings, lazy):
+        t0 = time.perf_counter()
+        codes = load_helper(data, label, timings, lazy)
+        exec_helper(codes, label, timings)
+        t4 = time.perf_counter()
         print(f"       {label} total: {t4-t0:.3f}")
         return timings

     code = compile(source, "<old>", "exec")
     data = marshal.dumps(code)
-    classic_timings = helper(data, "Classic", lazy=False)

-    new_timings = helper(data, "Lazy", lazy=True)
+    #new_timings = helper(data, "Lazy", new_timings, lazy=True)
+    #classic_timings = helper(data, "Classic", classic_timings, lazy=False)
+
+    new_timings = {}
+    classic_timings = {}
+    classic_codes = load_helper(data, "Classic", classic_timings, lazy=True)
+    exec_helper(classic_codes, "Classic", classic_timings)
+    lazy_codes = load_helper(data, "Lazy", new_timings, lazy=True)
+    exec_helper(lazy_codes, "Lazy", new_timings)

```

**Comment by gvanrossum** [comment]
> In lazy_marshal_perf.py, if I reverse the order so that we calculate new_timings first and classic_timings second, the result changes quite dramatically (classic exec becomes slower):

Very weird, I see this too. I've been in meetings all morning, hopefully I have time to look into this and the next one more later. But I should warn you that I'm going in vacation for 2+ weeks starting Friday, returning around August 8.

**Comment by gvanrossum** [comment]
> That loop is probably not the issue, it doesn't run after the first exec.

Hm, true.

**Comment by gvanrossum** [comment]
I give up.

**Comment by gvanrossum** [comment]
Okay, here's an interesting result. If I reuse the same dict object as the namespace for all 4000 exec() calls, clearing it each time before the call, the steady-state execs all get faster. Code like this:
```py
    def exec_helper(codes, label, timings):
        timings["execs"] = []
        ns = {}
        for i in range(4):
            t3 = time.perf_counter()
            for code in codes:
                ns.clear()
                exec(code, ns)
            t4 = time.perf_counter()
            print(f"{label} exec #{i+1}: {t4-t3:.3f}")
            timings["execs"].append(t4 - t3)
```
(Only lines using 'ns' are new/changed.)

Timing runs are still too noisy for me to see a significant different for the steady state.

**Comment by gvanrossum** [comment]
We may just be measuring GC performance. The classic load uses a ton of memory, all of which is released once the lazy load starts. The classic peaks at 2.9 Gb, while the lazy peaks at 0.73 Gb. (This is with the version of the script in the lazy-marshal branch, without #44 applied, but with your set of extreme parameters for unique-globals.)

**Comment by iritkatriel** [comment]
Right, but lazy load is what's making everything slow, not classic load.

**Comment by gvanrossum** [comment]
The effect we're seeing most strongly is that whichever set of execs we run
second is slower, though.

**Comment by iritkatriel** [comment]
This is what I'm talking about: https://github.com/faster-cpython/ideas/issues/66#issuecomment-884171248

The execs are in the same order, but moving the lazy load to the beginning makes a very big difference to the time of the classic load (which are first).

**Comment by iritkatriel** [comment]
Your result with the dict is interesting - clearing the dict probably breaks reference cycles and gets things GCed.

**Comment by gvanrossum** [comment]
> This is what I'm talking about: [#66 (comment)](https://github.com/faster-cpython/ideas/issues/66#issuecomment-884171248)
> 
> The execs are in the same order, but moving the lazy load to the beginning makes a very big difference to the time of the classic load (which are first).

We should probably think of this in terms of memory use.

Classic load: builds up 1000 deep trees, together 2.9 Gb.
Lazy load: builds up 1000 shallow trees (just dehydrated toplevels), very little memory used.
First lazy exec: hydrates 1000 toplevel code objects, together 0.7 Gb.
Subsequent lazy execs: no extra memory use.
Classic execs don't use extra memory either.

> Your result with the dict is interesting - clearing the dict probably breaks reference cycles and gets things GCed.

But just letting the dict being GG'ed (as exec(code, {}) does) should GC all that too -- when exec() returns there is nothing hanging on to the freshly created dict. But the next dict is probably freshly allocated at a different address, so maybe that flushes more L1 cache. (Can you check the dict code whether clear() resizes the hash table?)

**Comment by iritkatriel** [comment]
> > This is what I'm talking about: [#66 (comment)](https://github.com/faster-cpython/ideas/issues/66#issuecomment-884171248)
> > The execs are in the same order, but moving the lazy load to the beginning makes a very big difference to the time of the classic load (which are first).
> 
> We should probably think of this in terms of memory use.
> 
> Classic load: builds up 1000 deep trees, together 2.9 Gb.
> Lazy load: builds up 1000 shallow trees (just dehydrated toplevels), very little memory used.
> First lazy exec: hydrates 1000 toplevel code objects, together 0.7 Gb.
> Subsequent lazy execs: no extra memory use.
> Classic execs don't use extra memory either.

Then why does classic exec become slower if you do lazy load before it?

**Comment by iritkatriel** [comment]
> > Your result with the dict is interesting - clearing the dict probably breaks reference cycles and gets things GCed.
> 
> But just letting the dict being GG'ed (as exec(code, {}) does) should GC all that too -- when exec() returns there is nothing hanging on to the freshly created dict. But the next dict is probably freshly allocated at a different address, so maybe that flushes more L1 cache. (Can you check the dict code whether clear() resizes the hash table?)

Yes, it starts from scratch:

```
    d->ma_version_tag = DICT_NEXT_VERSION();
    dictkeys_incref(Py_EMPTY_KEYS);
    d->ma_keys = Py_EMPTY_KEYS;
    d->ma_values = empty_values;
```

I see in the debugger that ns is referred by the function objects it contains. Clearing the dict breaks that cycle, while just dropping the reference to it doesn't. This could help GC avoid cycle detection.

**Comment by gvanrossum** [comment]
> I see in the debugger that ns is referred by the function objects it contains.

Ah, that is for func_globals. Yeah, this is an unfortunate cycle.

> Clearing the dict breaks that cycle, while just dropping the reference to it doesn't.

Good point!

But how much of our other observations does this explain? I still see so much noise that I don't trust what I see on my machine.

**Comment by iritkatriel** [comment]
> But how much of our other observations does this explain? I still see so much noise that I don't trust what I see on my machine.

It makes me suspect that lazy load is creating a lot of objects that are reachable from this cycle, and which weigh on anything that comes after (including a classic exec).    And classic load doesn't do this.  So maybe we need to look for leaks.

**Comment by iritkatriel** [comment]
But if that was the issue, then this should have been as fast as with the dict.clear(), and it's not:

```
    def exec_helper(codes, label, timings):
        timings["execs"] = []
        for i in range(4):
            while gc.collect():   # <--------- added gc before timing
                pass
            t3 = time.perf_counter()
            for code in codes:
                exec(code, {})
            t4 = time.perf_counter()
            print(f"{label} exec #{i+1}: {t4-t3:.3f}")
            timings["execs"].append(t4 - t3)

```

**Comment by gvanrossum** [comment]
Hm, one thing that gets leaked is the bytes object (b) passed to marshal.loads(b, lazy=True). The context struct hangs on to this and that context struct is currently immortal. (This is pure laziness on my part, it should become an object so that it will be deleted when the last code object referring to it is hydrated *or deleted*.)

Hydrating also leaks the function name, qualname and filename -- again pure laziness.

**Comment by gvanrossum** [comment]
FWIW the context struct also hangs on to 'refs'.

**Comment by iritkatriel** [comment]
Ok this is weird. This is fast:

```
        for i in range(4):
            ns = {}
            t3 = time.perf_counter()
            for code in codes:
                exec(code, ns)
            t4 = time.perf_counter()
            print(f"{label} exec #{i+1}: {t4-t3:.3f}")
            timings["execs"].append(t4 - t3)
```

but change to `exec(code, {})` and it's slow.

**Comment by gvanrossum** [comment]
That overwrites the same keys in ns over and over, breaking most of the cycles, just as with ns.clear().

**Comment by iritkatriel** [comment]
no, I make a new ns for each iteration: `ns = {}`

**Comment by iritkatriel** [comment]
ah if I move the `ns = {}` into the inner loop then it's slow again.

**Comment by gvanrossum** [comment]
In my experience GC is weird. Try gc.disable(). That won't break any cycles, but if you manually break them (e.g. ns.clear()) garbage still gets cleared when the refcount goes to 0.

**Comment by iritkatriel** [comment]
With gc.disable() just before `t3 = ...` and gc.enable() + a gc loop just after `t4 = ...` (with the original `exec(code, {})`):

```
Lazy load: 0.270
Classic load: 4.474
Classic exec #1: 1.283
Classic exec #2: 1.290
Classic exec #3: 1.255
Classic exec #4: 1.244
Lazy exec #1: 3.453
Lazy exec #2: 1.205
Lazy exec #3: 1.249
Lazy exec #4: 1.236
Classic-to-new comparison:
>> load+exec ratio: 0.65 (new is 55% faster)
>> steady-state ratio: 0.99 (new is 1% faster)
```

These numbers make sense, and to me it confirms that the issue is lazy load trashing the cyclic GC.

**Comment by gvanrossum** [comment]
IIRC the gc module can report a bunch of stats. Maybe those provide some insight? Maybe gc.collect() returns how much it collected?

Or we can reason like this: 1000 separate loads create 1000 separate dehydrated objects, holding on to 1000 separate 'refs' lists (one each). The refs lists are probably long (to be confirmed via debugger) and each item is a reference to the toplevel dehydrated code object. Probably all those lists are being traced by the GC, but everything is still alive (because we have an outside reference to the topmost code object) and in addition everything there is being leaked (because the context struct is always leaked ATM).

So yeah, if you do things in that order (lazy load, classic load + exec, lazy exec) then the lazy data causes the GC to trace more stuff. (Not sure if it can be considered thrashing -- our GC is invoked every N allocations, and it's just less effective this way.)

**Comment by iritkatriel** [comment]
Why doesn't the classic exec time improve in the later iterations? Are these objects not being GCed?

**Comment by gvanrossum** [comment]
Code objects are not traversed by the GC, because they're immutable. So there's not much to traverse.

Lazy loading has those 1000 lists with lots of refs, which are all being traversed, pointlessly.

**Comment by iritkatriel** [comment]
GC stats:

Lazy load first:
```
Starting speed test: unique      globals        1000 funcs,   10 vars
Lazy load: 0.241
[{'collections': 36, 'collected': 117, 'uncollectable': 0}, {'collections': 2, 'collected': 28, 'uncollectable': 0}, {'collections': 2, 'collected': 10, 'uncollectable': 0}]
Classic load: 4.317
[{'collections': 1345, 'collected': 117, 'uncollectable': 0}, {'collections': 121, 'collected': 28, 'uncollectable': 0}, {'collections': 2, 'collected': 10, 'uncollectable': 0}]
Classic exec #1: 2.285
[{'collections': 2643, 'collected': 117, 'uncollectable': 0}, {'collections': 239, 'collected': 868896, 'uncollectable': 0}, {'collections': 13, 'collected': 120130, 'uncollectable': 0}]
Classic exec #2: 2.284
[{'collections': 3940, 'collected': 117, 'uncollectable': 0}, {'collections': 357, 'collected': 1745772, 'uncollectable': 0}, {'collections': 24, 'collected': 247257, 'uncollectable': 0}]
Classic exec #3: 2.252
[{'collections': 5239, 'collected': 117, 'uncollectable': 0}, {'collections': 475, 'collected': 2620646, 'uncollectable': 0}, {'collections': 34, 'collected': 365375, 'uncollectable': 0}]
Classic exec #4: 2.300
[{'collections': 6537, 'collected': 117, 'uncollectable': 0}, {'collections': 593, 'collected': 3496521, 'uncollectable': 0}, {'collections': 45, 'collected': 493503, 'uncollectable': 0}]
Lazy exec #1: 4.691
[{'collections': 7837, 'collected': 117, 'uncollectable': 0}, {'collections': 712, 'collected': 4380404, 'uncollectable': 0}, {'collections': 56, 'collected': 618628, 'uncollectable': 0}]
Lazy exec #2: 2.334
[{'collections': 9135, 'collected': 117, 'uncollectable': 0}, {'collections': 830, 'collected': 5255278, 'uncollectable': 0}, {'collections': 67, 'collected': 747757, 'uncollectable': 0}]
Lazy exec #3: 2.231
[{'collections': 10433, 'collected': 117, 'uncollectable': 0}, {'collections': 948, 'collected': 6130152, 'uncollectable': 0}, {'collections': 77, 'collected': 865875, 'uncollectable': 0}]
Lazy exec #4: 2.415
[{'collections': 11731, 'collected': 117, 'uncollectable': 0}, {'collections': 1066, 'collected': 7005026, 'uncollectable': 0}, {'collections': 88, 'collected': 995004, 'uncollectable': 0}]
Classic-to-new comparison:
>> load+exec ratio: 0.75 (new is 34% faster)
>> steady-state ratio: 1.05 (new is -5% faster)
```


Lazy load after classic exec:
```
Classic load: 4.455
[{'collections': 1343, 'collected': 117, 'uncollectable': 0}, {'collections': 121, 'collected': 28, 'uncollectable': 0}, {'collections': 2, 'collected': 10, 'uncollectable': 0}]
Classic exec #1: 1.646
[{'collections': 2641, 'collected': 117, 'uncollectable': 0}, {'collections': 239, 'collected': 869897, 'uncollectable': 0}, {'collections': 13, 'collected': 120130, 'uncollectable': 0}]
Classic exec #2: 1.636
[{'collections': 3939, 'collected': 117, 'uncollectable': 0}, {'collections': 357, 'collected': 1745772, 'uncollectable': 0}, {'collections': 24, 'collected': 248258, 'uncollectable': 0}]
Classic exec #3: 1.635
[{'collections': 5238, 'collected': 117, 'uncollectable': 0}, {'collections': 475, 'collected': 2621647, 'uncollectable': 0}, {'collections': 34, 'collected': 365375, 'uncollectable': 0}]
Classic exec #4: 1.624
[{'collections': 6536, 'collected': 117, 'uncollectable': 0}, {'collections': 593, 'collected': 3496521, 'uncollectable': 0}, {'collections': 45, 'collected': 494504, 'uncollectable': 0}]
Lazy load: 0.364
[{'collections': 6537, 'collected': 117, 'uncollectable': 0}, {'collections': 593, 'collected': 3496521, 'uncollectable': 0}, {'collections': 45, 'collected': 494504, 'uncollectable': 0}]
Lazy exec #1: 4.876
[{'collections': 7837, 'collected': 117, 'uncollectable': 0}, {'collections': 712, 'collected': 4376400, 'uncollectable': 0}, {'collections': 56, 'collected': 622632, 'uncollectable': 0}]
Lazy exec #2: 2.481
[{'collections': 9135, 'collected': 117, 'uncollectable': 0}, {'collections': 830, 'collected': 5251274, 'uncollectable': 0}, {'collections': 67, 'collected': 751761, 'uncollectable': 0}]
Lazy exec #3: 2.310
[{'collections': 10433, 'collected': 117, 'uncollectable': 0}, {'collections': 948, 'collected': 6128150, 'uncollectable': 0}, {'collections': 77, 'collected': 867877, 'uncollectable': 0}]
Lazy exec #4: 2.411
[{'collections': 11731, 'collected': 117, 'uncollectable': 0}, {'collections': 1066, 'collected': 7003024, 'uncollectable': 0}, {'collections': 88, 'collected': 997006, 'uncollectable': 0}]
Classic-to-new comparison:
>> load+exec ratio: 0.86 (new is 16% faster)
>> steady-state ratio: 1.48 (new is -33% faster)
```

**Comment by iritkatriel** [comment]
> Lazy loading has those 1000 lists with lots of refs, which are all being traversed, pointlessly.

Right, so we declare them immutable and make it stop looking at them?

**Comment by gvanrossum** [comment]
Hm, we'd have to change the data structure from a standard Python list to some custom flexible array. (Merely being immutable isn't enough -- even tuples must be traversed by GC, since a tuple could contain a list.)

Maybe we'd have to do that anyway, or maybe we'd even have to make code objects traversible (sadly, that would be slow) since the context struct makes things a bit iffy.

Is this a big issue for realistic code though? Our "benchmark" is pretty extreme.

**Comment by iritkatriel** [comment]
> Is this a big issue for realistic code though? Our "benchmark" is pretty extreme.

If I change it to shared variables we still seem to lose about 10%:

```
Classic load: 1.852
Classic exec #1: 0.746
Classic exec #2: 0.738
Classic exec #3: 0.723
Classic exec #4: 0.721
Lazy load: 0.028
Lazy exec #1: 1.229
Lazy exec #2: 0.793
Lazy exec #3: 0.823
Lazy exec #4: 0.800
Classic-to-new comparison:
>> load+exec ratio: 0.48 (new is 107% faster)
>> steady-state ratio: 1.11 (new is -10% faster)
```

**Comment by iritkatriel** [comment]
We should get to a place where a hydrated code object doesn't have any baggage from its lazy initialization, so at least when it's executed for the 10th time you can't tell the difference.

**Comment by iritkatriel** [comment]
Ah but these are not hydrated. Right. So we need to somehow make sure that GC doesn't have to deal with data attached to code objects that never execute.

**Comment by gvanrossum** [comment]
Looking at the 100 most popular PyPI packages (snapshot a few months old), we have nearly 10,000 files, with over 200,000 code objects, so about 20 code objects per file. Some big ones:
```
.\packages\pandas-1.3.0rc1.tar.gz                 :      1,270 files;     30,169 code objects;    501,579 lines
.\packages\matplotlib-3.4.2.tar.gz                :        881 files;     12,292 code objects;    226,178 lines
.\packages\scipy-1.7.0rc2.tar.gz                  :        821 files;     22,295 code objects;    393,249 lines
.\packages\SQLAlchemy-1.4.9.tar.gz                :        577 files;     34,091 code objects;    452,647 lines
```
Note that SQLAlchemy seems to have the most code objects per file.

**Comment by gvanrossum** [comment]
> We should get to a place where a hydrated code object doesn't have any baggage from its lazy initialization, so at least when it's executed for the 10th time you can't tell the difference.

After hydration we drop the context link, so that should be the case.

Since GC doesn't traverse code objects, they are not the problem -- but it finds all lists, and so the refs lists are a problem (not sure if they're *the* problem).

Realistically the refs lists and the bytes objects hanging off the context are immortal, unless you call every function and method in the module.

**Comment by gvanrossum** [comment]
> GC stats:

BTW what conclusion do you draw from that? (I find it hard to read since GitHub won't display all data at once even if I stretch the window.)

**Comment by iritkatriel** [comment]
The stats look the same in both cases, so it's consistent with what you say that there is work being done on traversing reachable stuff.

**Comment by iritkatriel** [comment]
> > We should get to a place where a hydrated code object doesn't have any baggage from its lazy initialization, so at least when it's executed for the 10th time you can't tell the difference.
> 
> After hydration we drop the context link, so that should be the case.
> 
> Since GC doesn't traverse code objects, they are not the problem -- but it finds all lists, and so the refs lists are a problem (not sure if they're _the_ problem).
> 
> Realistically the refs lists and the bytes objects hanging off the context are immortal, unless you call every function and method in the module.

What if the module managed the refs memory - it has a counter of how many dehydrated things it still has, when you hydrate something and this counter is 0 the junk is deleted. Would that work?

**Comment by gvanrossum** [comment]
I was thinking the refcount on the context object (once it is an object) would serve that purpose. I'd like to keep the module out of it since it's impossible to get to it from the code object, and from the module *to* the code object is tricky too.

**Comment by iritkatriel** [comment]
Ok, that works too. So it's just a matter of making them something the GC leaves alone (not lists). Then the context object is responsible for removing them.

**Comment by methane** [comment]
You can untrack list object by `_PyObject_GC_UNTRACK(refs)`.

**Comment by gvanrossum** [comment]
>  You can untrack list object by `_PyObject_GC_UNTRACK(refs)`.

Oh, interesting. Irit, maybe you can put that in somewhere (I'm guessing when the context struct is first initialized) and see if that also fixes the slowdown -- then we'll know that this was the cause!

**Comment by methane** [comment]
```
diff --git a/Python/marshal.c b/Python/marshal.c
index 363bcabf75..daf3eea2c2 100644
--- a/Python/marshal.c
+++ b/Python/marshal.c
@@ -1961,6 +1961,7 @@ marshal_loads_impl(PyObject *module, Py_buffer *bytes, int lazy)
         rf.ctx->len = n;
         rf.ctx->code = NULL;
         Py_INCREF(rf.refs);
+        PyObject_GC_UnTrack(rf.refs);
         rf.ctx->refs = rf.refs;
     }
     result = read_object(&rf);
```

**Comment by iritkatriel** [comment]
It's better with untrack. (I took the latest version and change the perf script back to `exec(code, {})`

```
Starting speed test: unique      globals        1000 funcs,   10 vars
Classic load: 4.456
Lazy load: 0.278
Classic exec #1: 1.834
Classic exec #2: 1.962
Classic exec #3: 1.905
Classic exec #4: 1.964
Lazy exec #1: 4.457
Lazy exec #2: 1.794
Lazy exec #3: 1.733
Lazy exec #4: 1.771
Classic-to-new comparison:
>> load+exec ratio: 0.75 (new is 33% faster)
>> steady-state ratio: 0.90 (new is 11% faster)
```

But there is probably more that we need to untrack. (Or maybe it will help to not leak these lists.) If I move the lazy load after the classic exec then the classic exec is even faster:

```
Starting speed test: unique      globals        1000 funcs,   10 vars
Classic load: 4.467
Classic exec #1: 1.660
Classic exec #2: 1.657
Classic exec #3: 1.598
Classic exec #4: 1.613
Lazy load: 0.336
Lazy exec #1: 4.121
Lazy exec #2: 1.674
Lazy exec #3: 1.654
Lazy exec #4: 1.694
Classic-to-new comparison:
>> load+exec ratio: 0.73 (new is 37% faster)
>> steady-state ratio: 1.05 (new is -5% faster)
```

**Comment by iritkatriel** [comment]
There are two other places where the refs list gets allocated. If I untrack there as well it looks pretty good:
```
Starting speed test: unique      globals        1000 funcs,   10 vars
Classic load: 4.716
Lazy load: 0.281
Classic exec #1: 1.721
Classic exec #2: 1.644
Classic exec #3: 1.680
Classic exec #4: 1.750
Lazy exec #1: 4.530
Lazy exec #2: 1.765
Lazy exec #3: 1.680
Lazy exec #4: 1.696
Classic-to-new comparison:
>> load+exec ratio: 0.75 (new is 34% faster)
>> steady-state ratio: 0.97 (new is 3% faster)
```


```
Starting speed test: unique      globals        1000 funcs,   10 vars
Classic load: 4.317
Classic exec #1: 1.718
Classic exec #2: 1.636
Classic exec #3: 1.620
Classic exec #4: 1.605
Lazy load: 0.359
Lazy exec #1: 4.036
Lazy exec #2: 1.693
Lazy exec #3: 1.648
Lazy exec #4: 1.663
Classic-to-new comparison:
>> load+exec ratio: 0.73 (new is 37% faster)
>> steady-state ratio: 1.04 (new is -3% faster)
```

**Comment by gvanrossum** [comment]
Okay, just send a PR! If you would like to work on the not leaking part,
please go ahead -- I'm packing for my trip today.

**Comment by iritkatriel** [comment]
I'm not sure that untracking the refs is ok. If a ref leads to another code object we can end up with an unreachable cycle that is never collected.

**Comment by methane** [comment]
> I'm not sure that untracking the refs is ok. If a ref leads to another code object we can end up with an unreachable cycle that is never collected.

If it is not OK, we need to visit refs. All code objects must be GC object. It is unacceptable.
So we need to avoid cycle reference. e.g. `code -> context -> refs -> code`:

Can we make it lazy only for leaf code objects?

**Comment by iritkatriel** [comment]
> All code objects must be GC object. It is unacceptable.

Aren't they already?

```
/* Bytecode object */
struct PyCodeObject {
    PyObject_HEAD
```

> Can we make it lazy only for leaf code objects?

I don't think that's relevant because the refs are not necessarily pointing to nested code objects. (It's also probably not feasible because you don't know it's a leaf before you hydrate it.)

But maybe the same reasoning about recursion here https://github.com/faster-cpython/ideas/issues/66#issuecomment-882876587 also applies to cycles, and it's actually fine.

I think we need to initially not untrack the refs, see what happens in real programs, and then do further analysis to determine what's safe.

**Comment by methane** [comment]
> Aren't they already?
> 
> ```
> /* Bytecode object */
> struct PyCodeObject {
>     PyObject_HEAD
> ```

It is not `PyObject_GC_HEAD`.
Python GC can not break circular reference including code object.
So tracking refs doesn't make sense at all.

**Comment by gvanrossum** [comment]
I think we're heading towards grooming this for merging upstream. But before that we should probably show some proof that it actually helps (so far I've mostly got hunches).

We also have some TODO items left (I'm starting a new list):

- [ ] Crashes. See below.
- [ ] Leaks in _PyCode_Update() -- it doesn't DECREF objects already in the code object.
- [ ] Use lazy marshal by default, with a -X flag to disable it (stop testing $LAZY).

**Comment by iritkatriel** [comment]
I'm trying to find an example of where this helps. A good one would be something with numpy/scipy/jupyter, but they are not compatible with 3.11 yet. I think we need something that uses only the stdlib. 

I can think of two types of cases:
1. interactive, where init time is noticable to a user.
2. Splitting a calculation on a grid (map-reduce type thing) where each grid task does the imports so that time is not amortized.

Is a unittest run a good candidate?

**Comment by iritkatriel** [comment]
> Is a unittest run a good candidate?

Probably not, because if we have good coverage then everything gets hydrated.

**Comment by gvanrossum** [comment]
Maybe a module like os.path, which has lots of utility functions, most of which aren't used by most programs. Or argparse, ditto.

**Comment by iritkatriel** [comment]
I don't think I can see a consistent difference in the time of just importing one of them with lazy vs not.  If you look at argparse, it has a number of imports inside functions, so it seems like it was optimised.

**Comment by iritkatriel** [comment]
For reference, here are some 3.9 import times:

```
iritkatriel@Irits-MBP cpython % time ./python.exe -c "import numpy"
./python.exe -c "import numpy"  1.47s user 0.14s system 151% cpu 1.058 total
iritkatriel@Irits-MBP cpython % time ./python.exe -c "import scipy"
./python.exe -c "import scipy"  0.99s user 0.12s system 236% cpu 0.468 total
iritkatriel@Irits-MBP cpython % time ./python.exe -c "import pandas"
./python.exe -c "import pandas"  1.76s user 0.17s system 157% cpu 1.224 total
iritkatriel@Irits-MBP cpython % time ./python.exe -c "import argparse"
./python.exe -c "import argparse"  0.15s user 0.01s system 97% cpu 0.169 total
iritkatriel@Irits-MBP cpython % time ./python.exe -c "import os.path" 
./python.exe -c "import os.path"  0.08s user 0.01s system 96% cpu 0.091 total
```

**Comment by iritkatriel** [comment]
I wonder how hard it would be to merge the lazy marshal code to 3.9. I'll try.

**Comment by gvanrossum** [comment]
Oh, that should be easy, marshal doesn’t change much (though code object fields have changed).

**Comment by iritkatriel** [comment]
Yes, I need to decide whether to backport the CodeObjectConstructor thing (probably not). The test looks easy.

**Comment by gvanrossum** [comment]
Don’t backport that, the marshal code was originally designed without that. Just copy all the code we need into the Update function.

**Comment by iritkatriel** [comment]
This is a 3.10 backport: https://github.com/iritkatriel/cpython/pull/23
With LAZY=1 it crashes on test_import and test_subprocess.

**Comment by gvanrossum** [comment]
Wow. So is it effective for numpy or pandas?

**Comment by iritkatriel** [comment]
It doesn't with with numpy I install with pip:

```
Python 3.10.0b4+ (heads/3.10-lazy-marshal:b2116e1158, Jul 30 2021, 19:12:59) [Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy
Assertion failed: (_PyCode_IsHydrated(code)), function _PyFrame_New_NoTrack, file Objects/frameobject.c, line 840.
zsh: abort      ./python.exe
```

(without LAZY=1 it does import in 3.10).

Might work if I build numpy from src (never done it before).  I wanted to also revisit the 3.9 backport - maybe backport the 3.10 to 3.9 because it's less messy and quite close.

**Comment by iritkatriel** [comment]
Actually I'll try to debug this. Might be just missing a Hydrate call somewhere.

**Comment by iritkatriel** [comment]
ok, with the last commit it does import. But either I'm doing something wrong or it's not making a difference:

On windows 315ms --> 307 ms, and on Mac:

iritkatriel@Irits-MBP cpython % time ./python.exe -c "import numpy"
./python.exe -c "import numpy"  0.42s user 0.03s system 98% cpu 0.451 total
iritkatriel@Irits-MBP cpython % export LAZY=1
iritkatriel@Irits-MBP cpython % time ./python.exe -c "import numpy"
./python.exe -c "import numpy"  0.39s user 0.03s system 99% cpu 0.419 total

I don't know why it's so fast now, compared to what I ran a few days ago. Maybe it's missing dependencies or something.

**Comment by iritkatriel** [comment]
I'm going to assume that 3.10 results are not reliable (maybe not everything is working... ) and proceed with the 3.9 backport.

**Comment by gvanrossum** [comment]
But why would 3.9 be any different?

Could PGO be the difference with a few days ago?

**Comment by iritkatriel** [comment]
You're asking as if you expect me to make sense. :)

With pgo on the mac it won't import numpy at all. Says "no muldule named 'numpy.core_multiarray_umath'."

**Comment by gvanrossum** [comment]
Maybe this is another dead end (like Experiment A) :-(.

**Comment by iritkatriel** [comment]
I'm not ready to conclude that. I don't believe the 3.10 numbers I got. Either I did something wrong or something is not working (the numpy test suite didn't pass either).

**Comment by gvanrossum** [comment]
Sure. I’m still on vacation and can’t really think about this too much. :-)

**Comment by iritkatriel** [comment]
Then focus on positive thoughts.  I'll try to get a 3.9 version working.

**Comment by iritkatriel** [comment]
The new 3.9 branch is at: https://github.com/iritkatriel/cpython/pull/25

two unit tests are still failing: test_frozen test_importlib 
However, the micro benchmarks look the same as in 3.11 (see https://github.com/iritkatriel/cpython/pull/25#issuecomment-890590802), and import numpy works (with pgo). The numpy test suite pretty much passed (14848 tests passed, 3 failed).

I added counters which are printed in the end how many times PyCode_New/PyCode_Udpate/Hydate are called. 
pandas/numpy import: https://github.com/iritkatriel/cpython/pull/25#issuecomment-890597046
We don't 
We don't gain much time, even thought that stats indicate that quite a lot is not being hydrated. Unless I'm counting wrong.

**Comment by iritkatriel** [comment]
I'm not sure my counters are the correct ones - PyCode_New doesn't mean that we created a dehydrated object. I'll look into it tomorrow.

**Comment by iritkatriel** [comment]
I updated the counter and it looks like we hydrate about half the code object we create. It's possible we could reduce that if we look at where we are hydrating and manage to avoid some cases, but I'm not sure that would help since not hydrating half didn't give a speedup.

**Comment by gvanrossum** [comment]
But what about code objects whose parent is never hydrated? Those aren't even counted.

Experiment B (#65) has some ideas for reducing sharing between code objects, it might help. Then again it might help without Experiment C.

**Comment by iritkatriel** [comment]
They are counted in the classic run. But anyway, if we're hydrating less than half and it shows no speed up that just makes it worse.

If you look at the microbenchmark results, they show that you only get a speed up if there are many locals. Maybe there aren't that many?

**Comment by gvanrossum** [comment]
Can we profile the marshal code to see where the most time is spent?

**Comment by iritkatriel** [comment]
> Can we profile the marshal code to see where the most time is spent?

That would be good. @ericsnowcurrently can you advise how?

**Comment by iritkatriel** [comment]
I got something out of the profiler that comes with xcode. I'll play with it tomorrow and try to come up with something.

**Comment by iritkatriel** [comment]
This seems to be a hotspot for the micro-benchmark  "unique      globals        1000 funcs, 1000 vars, 1000 reps".

<img width="942" alt="Screenshot 2021-08-04 at 00 53 02" src="https://user-images.githubusercontent.com/1055913/128100655-ab5eeeeb-0461-4ac4-8f44-4e4a3ff91a36.png">

**Comment by iritkatriel** [comment]
<img width="787" alt="Screenshot 2021-08-04 at 00 58 04" src="https://user-images.githubusercontent.com/1055913/128100922-181f4d7c-eec1-4128-b616-4fd336b754c9.png">

**Comment by iritkatriel** [comment]
I'm trying to see where this spends its time (with classic marshal):
(The loop is so that I can click play on the profiler after startup and see something meaningful.)

```
import sys
keep = set(sys.modules.keys())
while True:
  import numpy
  for m in list(sys.modules.keys()):
      if m not in keep:
          del sys.modules[m]
```


If I root the profile at the import (so the import is 100%), it eventually gets to

```
93.9%  PyImport_ImportModuleLevelObject
which splits into
51.5% --> a call path which later splits into 
    27.7% --> marshal_loads
    22.0% --> PyEval_EvalCode which seems to spend most of its time calling builtin_compile 
41.9% --> import_find_and_load
    which spends a lot of time in the compiler/ast area as well. I don't think marshal shows up there at all.
```

So I think marshal is only part of the 27%.

**Comment by iritkatriel** [comment]
Inside marshal, it looks like a lot of time is spent in GC.

<img width="628" alt="Screenshot 2021-08-04 at 15 49 55" src="https://user-images.githubusercontent.com/1055913/128202826-d7bf33a3-b60e-4fba-859f-761a072628fe.png">

**Comment by gvanrossum** [comment]
May have to take a look at the code to understand what it is *compiling*. Because I didn’t expect it to.

The gc thing sounds like it’s using a lot of memory — every N allocations it tries to run gc. Which makes sense since it’s numpy.

**Comment by iritkatriel** [comment]
Note that these results are from a debug build. With an optimised one I can't see as much. Marshal seems even less significant.

**Comment by iritkatriel** [comment]
I think it's things like this
https://github.com/numpy/numpy/blob/ecdba3a140126174a6a4d86fe0fbf2ce2ad680f9/numpy/core/overrides.py#L145

called from places like
https://github.com/numpy/numpy/blob/ecdba3a140126174a6a4d86fe0fbf2ce2ad680f9/numpy/fft/_pocketfft.py#L122

**Comment by iritkatriel** [comment]
If we add an "overrides dict" arg to functool.wraps so you can specify fields explicitly then they could do what they need to do without compiling.

**Comment by gvanrossum** [comment]
> If we add an "overrides dict" arg to functool.wraps so you can specify fields explicitly then they could do what they need to do without compiling.

Wow. And `functools.wraps` is pretty trivial to write in pure Python (or it would be if it wasn't written to be overly general :-).

**Comment by iritkatriel** [comment]
It's just this one place, I didn't find any more.  This is the commit where it was introduced: https://github.com/iritkatriel/numpy/commit/34bca67c9c918e90db4b33336a1d9ed2a57350d4

**Comment by iritkatriel** [comment]
> Wow. And `functools.wraps` is pretty trivial to write in pure Python (or it would be if it wasn't written to be overly general :-).

`functools.wraps`  is pure-python in functools, it just uses partial from _functools.

**Comment by iritkatriel** [comment]
Actually is there more to what they're doing that this?

```
        @functools.wraps(implementation)
        def public_api(*args, **kwargs):
             relevant_args = dispatcher(*args, **kwargs)
             return implement_array_function(
                implementation, public_api, relevant_args, args, kwargs)

        public_api.__name__ = implementation.__name__
```

**Comment by iritkatriel** [comment]
I think this might work actually:

```
        @functools.wraps(implementation)
        def public_api(*args, **kwargs):
             public_api.__name__ = implementation.__name__     ##### ADDED THIS
             relevant_args = dispatcher(*args, **kwargs)
             return implement_array_function(
                implementation, public_api, relevant_args, args, kwargs)

        public_api.__name__ = implementation.__name__

```

**Comment by gvanrossum** [comment]
I'm back, though due to jet lag my brain is only at 50% capacity. :-( This morning we had a discussion about startup time (again) and it looks like we still don't know how much time is spent where. I am trying to understand what's going on in https://github.com/faster-cpython/ideas/issues/75#issuecomment-887677654, which sounds like related to your observation about marshal spending a lot of time in GC a few comments up...

**Comment by iritkatriel** [comment]
Welcome back.  I made the change above to numpy, and it sped up import by 10% on a debug version of python (though didn't make much difference on an optimised version, not sure why).   Also, pandas/scipy/tensorflow all import numpy so we won't make them fast without solving numpy.

We should probably tell the numpy people about this slow compile, but I wanted you to have a second look at this before I write them and say something dumb.

Yes, GC is also interesting.

**Comment by gvanrossum** [comment]
> I think this might work actually:
> 
> ```
>         @functools.wraps(implementation)
>         def public_api(*args, **kwargs):
>              public_api.__name__ = implementation.__name__     ##### ADDED THIS
>              relevant_args = dispatcher(*args, **kwargs)
>              return implement_array_function(
>                 implementation, public_api, relevant_args, args, kwargs)
> 
>         public_api.__name__ = implementation.__name__
> ```

What's going on here? Why is the wraps() call not setting `public_api.__name__ = implementation.__name__`? And why do you set it manually twice, once outside and once inside the function? Is `implementation.__name__` dynamic?

**Comment by iritkatriel** [comment]
So this is to replace
https://github.com/numpy/numpy/blob/ecdba3a140126174a6a4d86fe0fbf2ce2ad680f9/numpy/core/overrides.py#L136

I'm guessing they did that because implement_array_function expects the second arg to be a function with name of implementation.__name__ rather than public_api.  My last line is indeed probably not necessary.

**Comment by gvanrossum** [comment]
I saw the numpy commit that introduced that (you linked earlier to it: https://github.com/iritkatriel/numpy/commit/34bca67c9c918e90db4b33336a1d9ed2a57350d4) but it said that it was to make prettier tracebacks. The only reason I can think of that wraps() doesn't have the desired effect is that the traceback code gets the function name from the code object, not from the function object. Since the code object is immutable, wraps() has no power over it.

(Did you verify the tracebacks of the examples from the commit after your proposed fix?)

Perhaps we can use code.replace() instead? E.g.
```
public_api.__code__ = public_api.__code__.replace(co_name=implementation.__name__)
```

**Comment by iritkatriel** [comment]
Ah I see, no I wrote my own test which was irrelevant to this. 

It gives those tracebacks if we do:

```
        @functools.wraps(implementation)
        def public_api(*args, **kwargs):
            relevant_args = dispatcher(*args, **kwargs)
            return implement_array_function(
                implementation, public_api, relevant_args, args, kwargs)
 
        public_api.__code__ = public_api.__code__.replace(
                                        co_name=implementation.__name__,
                                        co_filename='<string>')

```

**Comment by gvanrossum** [comment]
I don't see why we need to replace co_filename with `<string>`. (It's merely an artefact of the compile() call.) I think we might have to add co_qualname, at least in 3.11.

**Comment by iritkatriel** [comment]
I don't know what it should be, our point now is to show the numpy people that they can do what they are doing without compiling. They can set the filename to whatever they want.

**Comment by gvanrossum** [comment]
Yeah, tyou can go ahead and get their attention to this issue now.

**Comment by iritkatriel** [comment]
Done: https://github.com/numpy/numpy/pull/19638

**Comment by gvanrossum** [comment]
Thanks! Bikeshedding has already started on the PR so that's good news.

Now let's see if we can't get some numbers proving that our approach actually is a win. (My intuition is that it's actually the biggest win for packages that haven't been optimized to deal with slow imports yet, not for startup. Though Experiment B would benefit that.)

**Comment by gvanrossum** [comment]
Here are some numbers (but not enough). I looked into tkinter. I instrumented marshal.c so that it increments one of two counters at the calls to `_PyCode_Update()` and `_PyCode_New()`, respectively. (This should catch all code objects created or updated by marshal.loads().)

- Calling just `marshal.loads()` on the file `Lib\\tkinter\\__pycache__\\__init__.cpython-311.pyc` creates 224 code objects and of those, 18 are hydrated (presumably by the loop in `case TYPE_REF:`, due to constant sharing).
- After `import tkinter` (to warm up some things), calling `importlib.reload(tkinter)` creates a total of 559 code objects, and of those, 110 are hydrated. These numbers include the 224/18 of above; the difference ought to be explainable by the toplevel code object and the code objects representing classes being executed and the nested code objects representing methods being materialized (though remaining dehydrated), and some hydration (mostly due to constant sharing, as there's hardly any initialization code in this module). I found 41 classes and 511 functions/methods, so there's still a little difference, but the order of magnitude is correct.
- In non-lazy mode, I count a total of 569 code objects. I think the 10 extras are comprehensions -- since no methods are called, the methods' code objects aren't hydrated, so the nested code objects for those aren't even created in hydrated form. (There's also one lambda, for which the same applies.)

Of these numbers, the 224 code objects created for just marshal.loads() (which executes no code) is perhaps the most disturbing. I would have expected a single code object, the toplevel code for the module, in dehydrated state. What am I missing?

**Comment by gvanrossum** [comment]
Well, okay, this is interesting. The toplevel code object (and probably others) is written with `flag==0x80` due to my fix for shared code objects (98f4315ae3e8c5ae98e83280e93d0af94f273407, literally the last commit in my lazy-marshal branch).

Maybe we'll get better results if we roll that back? Or perhaps it's a red herring, since only the toplevel code object should be shared, and it will be hydrated as soon as we execute the module.

For now I think this is a red herring. But I still haven't gotten any further evidence that we're saving anybody time with lazy marshalling.

**Comment by gvanrossum** [comment]
So here's a further hurdle towards measuring our profits. I created a simpler test (simpler than lazy_marshal_perf.py) that writes 100 tiny functions to a module and then measures a few things:
- Time to read the file from disk in memory
- Time to unmarshal the data in the file
- Time to execute the file

Normalized to a cost **per function** I find:
- I/O takes 3-5 usec per func
- marshal.loads takes 0.3-0.5 usec per func
- creating a function object costs 0.1 usec per func
- the whole import takes ~10-12 usec per func

Half the time is unaccounted for here; I presume this is executing the bytecode of the toplevel module (100x LOAD_CONST, MAKE_FUNCTION, STORE_NAME).

But the conclusion is that for this case, speeding up marshal won't help. :-(

**Comment by gvanrossum** [comment]
Note that this is on Windows, which may explain the dreadful I/O time. --
--Guido (mobile)

**Comment by iritkatriel** [comment]
If I import Tkinter 1000 times, it seems that lazy is, if anything, slightly slower at the moment. So yeah, it doesn't seem to be helping.

```
import sys
keep = set(sys.modules.keys())
for i in range(1000):
  import tkinter
  for m in list(sys.modules.keys()):
      if m not in keep:
          del sys.modules[m]
```

% export LAZY=1
% time ./python.exe t.py
lazy = 1
./python.exe t.py  7.59s user 1.15s system 99% cpu 8.761 total
% time ./python.exe t.py
lazy = 1
./python.exe t.py  7.57s user 1.12s system 99% cpu 8.708 total
% unset LAZY
% time ./python.exe t.py
lazy = 0
./python.exe t.py  7.45s user 0.77s system 99% cpu 8.220 total
% time ./python.exe t.py
lazy = 0
./python.exe t.py  7.56s user 0.80s system 99% cpu 8.379 total

**Comment by gvanrossum** [comment]
So on our benchmark machine, overall is 1.1 usec/func, I/O only 0.1 usec/func, unmarshal 0.3-0.4 usec/func independent of $LAZY, function creation is 0.1 usec/func. (No PGO, non-debug build; same as before.)

This suggests the machine is about as fast as my Windows box (I guess it's the same CPU just more cores?) but I/O is blindingly faster (no surprise there, though it suggests that freezing modules will help with startup time). Some other overhead seems less to, since the overall import time is about 10x faster.

The marshal overhead is about the same with and without lazy, which matches your latest finding.

We do see that marshal is around a third of the total import time here. Perhaps we simply need to look to make the construction of a dehydrated code object from the marshal data faster? Maybe a profiler can tell us what takes the most time in the `case TYPE_CODE:` block. I wonder if inlining `r_long()` would help (or at least the fast path for loads). But we need to know where the time goes first (it's possible that constructing the code object takes more time, for example).

**Comment by iritkatriel** [comment]
Maybe the r_ref_reserve loop plus the r_ref_insert loop add up to what we save?

**Comment by gvanrossum** [comment]
Possibly (though I'm skeptical that it's that much).

Mark pointed out (and I've since verified) that the I/O overhead is largely a constant per file, not per func.

**Comment by gvanrossum** [comment]
Yeah, that loop feels a bit suspicious. But I think I'm personally going to give this some rest. I'll be writing up a new proposal shortly. In the meantime, are you at all interested in Experiment B? It's pretty simple really.

**Comment by iritkatriel** [comment]
> In the meantime, are you at all interested in Experiment B? It's pretty simple really.

Yes, I'll have a look.

**Comment by gvanrossum** [comment]
Closing, we decided to go a different way (deep-freeze).
